### PR TITLE
[Button Block]: Add padding block support

### PR DESCRIPTION
**Pull request by stacimc**
## Description

Adds padding support to the Button block using the spacing block support.
Issue #31731

**Notes:**
The block support currently allows arbitrarily large padding values which can cause blocks implementing it to overflow the page. See #31773

## How has this been tested?

Manually.

**Test Setup**
Your theme will need to opt into padding support. You can do this by updating your `theme.json` to enable padding under the spacing control, in either your default settings or the block context for button:
```
"settings": {
    "spacing": {
        "customPadding": true,
        "units": [
            "px",
            "em",
            "rem",
            "vh",
            "vw"
        ]
    }
}
```

**Test Instructions**

1. Create a post, add a buttons block with a button.
2. Select the button block and confirm padding controls in the sidebar work, adjusting units/padding per side.
3. Add another button and confirm that its padding is default. Update its padding and verify that it does not affect the padding of the first button.
3. Save the post and confirm the padding is visually correct on the frontend.

## Screenshots 
Screen Capture on 2021-05-27 at 13-55-29

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. 
- [ ] My code follows the accessibility standards. 
- [ ] I've tested my changes with keyboard and screen readers. 
- [ ] My code has proper inline documentation. 
- [ ] I've included developer documentation if appropriate. 
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal).

### CODE REVIEW

**Comment by ramonjd** [comment]
This is testing really well. ✅ 

button-padding


Editor | Frontend
------------ | -------------
Screen Shot 2021-05-17 at 11 28 37 am | Screen Shot 2021-05-17 at 11 28 40 am



I think this PR requires a rebase to fix the failing PHP lint issues.

File: packages/block-library/src/button/edit.js
Comment (ramonjd) [inline packages/block-library/src/button/edit.js:236]: Should this be ported to [edit.native.js](https://github.com/WordPress/gutenberg/blob/e85cfce55009f01c223faa149ad5f19e5f29c29e/packages/block-library/src/button/edit.native.js#L454) as well? Asking because I don't know :) and I can't yet get the native mobile test env running locally.

File: packages/block-library/src/button/save.js
Comment (ramonjd) [inline packages/block-library/src/button/save.js:14]: The `getSpacingClassesAndStyles` import is causing errors in the Android native app. 

Screen Shot 2021-05-20 at 1 31 35 pm


We also have to import `getSpacingClassesAndStyles` for native in https://github.com/WordPress/gutenberg/blob/e66e6d98d70b5a339f4548e67872122024e761e4/packages/block-editor/src/hooks/index.native.js

`export { getSpacingClassesAndStyles } from './use-spacing-props';`

File: packages/block-editor/src/index.js
Comment (ramonjd) [inline packages/block-editor/src/index.js:11]: We also have to import this for native in https://github.com/WordPress/gutenberg/blob/e66e6d98d70b5a339f4548e67872122024e761e4/packages/block-editor/src/hooks/index.native.js

`export { getSpacingClassesAndStyles } from './use-spacing-props';`

File: packages/block-library/src/button/edit.js
Comment (ramonjd) [reply packages/block-library/src/button/edit.js:236]: So I got it running and saw that we do have button props. Color and spacing aren't there. Maybe for a follow up PR?

Screen Shot 2021-05-20 at 1 54 04 pm

**Review by ramonjd — CHANGES_REQUESTED** [review]
This is looking great. 

We need to import the new hook `getSpacingClassesAndStyles` in [index.native.js](https://github.com/WordPress/gutenberg/blob/e66e6d98d70b5a339f4548e67872122024e761e4/packages/block-editor/src/hooks/index.native.js#L13)

Since this PR doesn't add the button padding controls to native there's no change. Should we add it in another PR?

**Comment by stacimc** [comment]
> This is looking great.
> 
> We need to import the new hook `getSpacingClassesAndStyles` in [index.native.js](https://github.com/WordPress/gutenberg/blob/e66e6d98d70b5a339f4548e67872122024e761e4/packages/block-editor/src/hooks/index.native.js#L13)
> 
> Since this PR doesn't add the button padding controls to native there's no change. Should we add it in another PR?

Thanks for the thorough testing @ramonjd -- I think the controls should be added to native in this PR as well. I'll get that added!

**Comment by stacimc** [comment]
I took a stab at adding native support for the button padding here: #32218. Until the `BoxControl` is supported natively, it won't be possible to match the UI exactly without essentially re-implementing that component. There's more information on that PR.

I updated this PR with the native exports so that it does not cause errors in the app, but think it might be better to continue work on the mobile side in a separate PR.

**Review by ramonjd — APPROVED** [review]
Nice!

I retested this PR on desktop and native mobile.

Desktop button padding support works as described in the test steps:
Screen Shot 2021-05-28 at 2 36 12 pm

Buttons controls on native work as expected in the editor and frontend after controls extraction:



Screen Shot 2021-05-28 at 2 41 49 pm



> I updated this PR with the native exports so that it does not cause errors in the app, but think it might be better to continue work on the mobile side in a separate PR.

Good point. 👍 Will be easier to test in a new PR too.

**Comment by jasmussen** [comment]
Thanks for working on this. Per the improvements that @aaronrobertshaw has been exploring with regards to a specific "Dimensions" panel, can we leverage some of that work for this one? See also https://github.com/WordPress/gutenberg/pull/32499 for an example of that panel.

**Comment by aaronrobertshaw** [comment]
> can we leverage some of that work for this one?

The changes in https://github.com/WordPress/gutenberg/pull/32392 will be picked up by all blocks using spacing support once it lands.

The "Spacing" panel and padding control shown in the gif in the PR description are injected by block supports. https://github.com/WordPress/gutenberg/pull/32392 updates the panel injected by the spacing support to use the new progressive disclosure block support panel.

**Review by aaronrobertshaw — COMMENTED** [review]
Nice work @stacimc ✨ 

This tests well for me.

- Individual button padding gets updated appropriately in the block editor and frontend
- Setting padding for buttons through the Global Styles works
- Theme.json generated styles also get correctly applied to the button both in editor and frontend

It might be worth raising here that there is a [discussion](https://github.com/WordPress/gutenberg/issues/33255) around how we can allow blocks to possibly apply styles from block supports to inner elements such as table rows or nav menus. 

If a solution to that is found it might allow the skipping of spacing support serialization in this block to be removed. I don't believe it should block this however.

Before merging we'll need to rebase this PR and get some design feedback that we can land this under the provision that the width and spacing panels will ultimately be unified once block support is updated to provide only a dimensions panel.

Being able to control button padding appears pretty important to me and really helps out block pattern designs. 

cc/ @jasmussen Any objections to merging this now?

**Comment by ramonjd** [comment]
I retested this today and it all looks good in the editor and frontends. Also rechecked on iOS.

button-padding

I can't see why we couldn't merge this one and wait for the Dimensions UI panel changes to be merged. As I understand it the Dimensions panel only updates the UI and doesn't affect this block's attributes, or requires writing deprecations.

Following @aaronrobertshaw's advice and pinging @geriux just to make sure the changes to the web button has no negative effects on native buttons.

**Comment by geriux** [comment]
> Following @aaronrobertshaw's advice and pinging @geriux just to make sure the changes to the web button has no negative effects on native buttons.

Thanks for the ping! I'll check it out 👍

**Comment by geriux** [comment]
I tested this by creating a post on the web editor with a button block and custom padding (following this PR testing instructions) and there are no negative effects while opening/modifying it on the mobile editor 👍

**Comment by aaronrobertshaw** [comment]
I've given this another quick double check as well. LGTM.

File: packages/block-library/src/button/edit.js
Comment (youknowriad) [inline packages/block-library/src/button/edit.js:203]: Can you clarify why this is needed here? Ideally padding support is enabled for blocks by just adding a block support flag without touching the block itself at all.

Comment (youknowriad) [reply packages/block-library/src/button/edit.js:203]: I guess this is because of the extra wrapper of the button block? That wrapper is hurting us a lot, ideally we should find ways to remove it without breaking changes. We tried this in the past but reverted. I wonder if we could make it opt-in somehow.

It seems for now we're just accepting the less than ideal button markup.

 cc @mtias

Comment (mtias) [reply packages/block-library/src/button/edit.js:203]: Definitely worth considering and exploring. I'm not too sure of what was the case with the extra wrapper.

**Comment by mtias** [comment]
@stacimc @ramonjd can we try disabling the granular padding for each edge and only keep general padding and/or axial padding and see if there are any concrete requests for the other?

**Comment by ramonjd** [comment]
> can we try disabling the granular padding for each edge and only keep general padding and/or axial padding and see if there are any concrete requests for the other?

I've been chatting to @aaronrobertshaw about this, and working out the best way to represent various configurations in the block supports JSON and/or what we need to do to get the existing control hooks to recognize them to display axial padding controls.

I think support for axial spacing can be achieved on top of work he's already done in https://github.com/WordPress/gutenberg/pull/30607 and by @andrewserong in https://github.com/WordPress/gutenberg/pull/32610.

I'll take a closer look. 👍 

I can't find any desperate pleas for allowing padding controls on individual edges so far. Still it sounds reasonable to able to specify between axial/individal and so on.

**Pull request merged**
